servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/RequestResponseCloseHandler.java

```diff
@@ -359,12 +399,49 @@ private void setSocketResetOnClose(final Channel channel) {
         }
     }
 
+    private void serverCloseGracefully(final Channel channel) {
+        // Perform half-closure as described in https://tools.ietf.org/html/rfc7230#section-6.6
+        serverHalfCloseInbound(channel);
+        serverHalfCloseOutbound(channel);
+    }
+
     private void serverHalfCloseInbound(final Channel channel) {
         assert !isClient;
-        if (!has(state, IN_CLOSED) && channel instanceof DuplexChannel) {
-            LOGGER.debug("{} Half-Closing INBOUND", channel);
+        if (!hasAny(state, DISCARDING_SERVER_INPUT, IN_CLOSED)) {
+            // Instead of actual half-closure DuplexChannel.shutdownInput() we discard all further inbound data, but
+            // keep reading to receive FIN from the remote peer.
+            LOGGER.debug("{} Discarding further INBOUND", channel);
             state = unset(state, READ);
-            ((DuplexChannel) channel).shutdownInput().addListener((ChannelFutureListener) this::onHalfClosed);
+            channel.pipeline().fireUserEventTriggered(DiscardFurtherInboundEvent.INSTANCE);
```

Comment:
`IN_SHUTDOWN` may be a bit misleading, because it actually is not shutdown. `IN_CLOSED` is set when the input is shutdown. 

The discard of input happens not only for graceful closure but also for other events, like protocol closing inbound/outbound, when we are not expecting more requests. Named it `DiscardFurtherInboundEvent` because it's kind of a request to discard without saying how exactly it should be discarded.

Happy to rename, just looking for better alternatives...